xline-client/tests/lock.rs

```diff
@@ -0,0 +1,83 @@
+use std::time::Duration;
+
+use common::get_cluster_client;
+use test_macros::abort_on_panic;
+use xline_client::{
+    error::Result,
+    types::lock::{LockRequest, UnlockRequest},
+};
+
+mod common;
+
+#[tokio::test]
+async fn lock_unlock_should_success_in_normal_path() -> Result<()> {
+    let (_cluster, client) = get_cluster_client().await?;
+    let client = client.lock_client();
+
+    let resp = client
+        .lock(LockRequest::new().with_name("lock-test"))
+        .await?;
+    assert!(resp.key.starts_with(b"lock-test/"));
+
+    client
+        .unlock(UnlockRequest::new().with_key(resp.key))
+        .await?;
+    Ok(())
+}
+
+#[tokio::test]
+#[abort_on_panic]
+async fn lock_contention_should_occur_when_acquire_by_two() -> Result<()> {
+    let (_cluster, client) = get_cluster_client().await?;
+    let client = client.lock_client();
+    let client_c = client.clone();
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+
+    tokio::spawn(async move {
+        let resp = client_c
+            .lock(LockRequest::new().with_name("lock-test"))
+            .await
+            .unwrap();
+        let _ignore = tx.send(());
+        tokio::time::sleep(Duration::from_secs(2)).await;
+        let _resp = client_c
+            .unlock(UnlockRequest::new().with_key(resp.key))
+            .await
+            .unwrap();
+    });
+
+    rx.recv().await.unwrap();
+    let now = tokio::time::Instant::now();
+    let resp = client
+        .lock(LockRequest::new().with_name("lock-test"))
+        .await
+        .unwrap();
+
+    assert!(now.elapsed() > Duration::from_secs(1));
```

Comment:
Why is `Duration::from_secs(1)` instead of `Duration::from_secs(2)`?

Reply:
The actual wait time will be around 2secs(more or less). 1s is a rather conservative value, just to demonstrate the lock request will not return immediately.

Reply:
How about this?
```suggestion
    let res = tokio::timeout(Duration::from_secs(2), client.lock(LockRequest::new().with_name("lock-test")))
    assert!(res.is_err());
    let resp = client
        .lock(LockRequest::new().with_name("lock-test"))
        .await
        .unwrap();
    assert!(resp.key.starts_with(b"lock-test/"));
```

Reply:
The duration may still need to be less than 2, because the sleep in another task may already began for a while before we wait for the timeout.

Reply:
This testing way lacks persuasiveness. Or, you can try a different approach.
```suggestion
    let (tx, rx) = std::sync::mpsc::sync_channel(0);  // rendezvous channel

    tokio::spawn(async move {
        let resp = client_c
            .lock(LockRequest::new().with_name("lock-test"))
            .await
            .unwrap();
        rx.recv().await.unwrap();
        let _resp = client_c
            .unlock(UnlockRequest::new().with_key(resp.key))
            .await
            .unwrap();
    });

    let res = tokio::timeout(Duration::from_secs(2), client.lock(LockRequest::new().with_name("lock-test")))
    assert!(res.is_err());
    tx.send();
    let res = tokio::timeout(Duration::from_micros(100), client.lock(LockRequest::new().with_name("lock-test")))
    assert!(res.is_ok());
    ….
```

Reply:
After change to this timeout syntax, I encounter an issue that the final lock request will never return. I find out that it may be because the lock is not cancel safe, after it canceled by timeout, it will not delete the key associated with the lock thus another lock attemp will just wait forever.

Reply:
after implemented cancellation clean up, the test is refactored base on your approach